TrueCrypt fixes related to keyfiles

AlekseyCherepanov opened the pull request:
`truecrypt2john.py` got simple fix. Without the fix, the tool omits keyfiles when only one keyfile is specified on command line.

Formats got fix for #5072. Each ciphertext with keyfiles gets its own allocated memory that the keyfiles are read into. So I can crack both volumes in the example with `TrueCrypt/TC-Volume`.

I used `mem_alloc_tiny`, so allocated memory is freed at the end of john's work. The proper way is to use `dyna_salt`, but I did not check this way yet. I had idea to collect allocated pointers into a list and free them in `done()`, but it seems similar to `mem_alloc_tiny`.

I repeated the fix for opencl format. But I did not test it. It turned out that keyfiles do not work at all (on pocl).

Cpu format is a bit tested.

So at the moment, this PR is a preview of changes.

Is it normal to include pointers into the salt struct? Comments mention salt dupe-removal. So I know that I touch ground unknown to me.

AlekseyCherepanov committed:
truecrypt2john.py: fixed check missing single keyfile

AlekseyCherepanov committed:
TrueCrypt cpu formats: convert keyfiles into kpool during loading
Partial fix for #5072

AlekseyCherepanov commented:
> Is it normal to include pointers into the salt struct? Comments mention salt dupe-removal. So I know that I touch ground unknown to me.

No, it is not normal. Example: the same ciphertext repeated twice gets loaded as `2 password hashes with 2 different salts`. (I guess pointers might be ok if they are same for same data.)

AlekseyCherepanov commented:
I implemented `dyna_salt` approach. It uses old global variable as temporary buffer, so I rewrote commit. Also I removed changes to opencl format.

Interestingly repeated ciphertext emitted message `Loaded 2 password hashes with no different salts`. After moving onto `dyna_salt`, I get this message: `Loaded 1 password hash`. So I changed path to a keyfile to include additional `../john-samples/` and I got again `Loaded 2 password hashes with no different salts`.

I did not test it fully yet.

`get_salt` with `dyna_salt` returns pointer to static buffer with pointer to actual structure. So I chose to make static variable with pointer to structure and return pointer to the variable without explicit buffer.

Code in `pkzip.c` for the same purpose:
```c
void *winzip_common_get_salt(char *ciphertext)
{
	uint64_t i;
	winzip_salt salt, *psalt;
	static unsigned char *ptr;
[...]
	if (!ptr)
		ptr = mem_alloc_tiny(sizeof(winzip_salt*),sizeof(winzip_salt*));
[...]
	memcpy(ptr, &psalt, sizeof(winzip_salt*));
	return (void*)ptr;
}
```

My code:
```c
static void* get_salt(char *ciphertext)
{
	static struct cust_salt salt, *s;
[...]
	s = &salt; /* assign every time */
[...]
	return &s;
}
```

Code in `pkzip.c` is more verbose. I dislike this verbosity. OTOH it allowed me to understand intention with buffer. Thinking in such terms, my code requires more steps in head to realize what is going on. So I am not sure that my code is better.

`assign every time` is an important point. I assign pointer to heap buffer to `s` later. With `s = &salt` in initialization of the static variable, the code was wrong.

I mix static and dynamic salts and everything works fine. At least ASan does not report leaks or attempts to `free()` static memory.

I think order of keyfiles does not affect result, so after more experiments I am going to add sorting to allow dupe-removal for such ciphertexts.

AlekseyCherepanov commented:
`pkzip.h`:
```c
typedef struct winzip_salt_t {
[...]
	unsigned char datablob[1];
} winzip_salt;
```

I use explicit flexible array member for `keyfiles_length`, i.e. I don't specify size. This feature is a part of C99. Hm, [Wikipedia article](https://en.wikipedia.org/wiki/Flexible_array_member) suggests that my style is not the best. Hm, I guess I need to review my use of this bit too: `s->dsalt.salt_cmp_offset = sizeof(s->dsalt);`

BTW I guess `pkzip.c` allocates 1 byte more than needed for salt.
```c
	psalt = (winzip_salt*)mem_calloc(1, sizeof(winzip_salt) + salt.comp_len);
```

But I need more than just one array: I have array `keyfiles_length` with variable length and then there is data from keyfiles that I store after the structure. With fixed length for `keyfiles_length`, the data could be part of structure. It would mean addition of 256 `int`s per salt.

AlekseyCherepanov commented:
`apply_keyfiles()` computes `kpool` from keyfiles and then mixes it into candidate. `kpool` can be computed loading ciphertexts. So `kpool` is 64 bytes big and it can replace all keyfiles.

`kpool` could be precomputed by `truecrypt2john.py` and embedded into ciphertext instead of filenames.

magnumripper commented:
I am unable to context-switch into this now but I did review your current PR and saw nothing that struck me as incorrect.

AlekseyCherepanov commented:
JFYI I found possible problem in my current code.

```c
static struct cust_salt {
[...]
	// Tightly packed blob with keyfiles' data is put right after the end.
	int keyfiles_length[];
} *psalt;
```

I expect such layout:
```
[main struct][keyfiles_length][blob with keyfiles]
             ^                ^ sizeof() + actual size of keyfiles_length[]
              \ sizeof(struct cust_salt)
```

The [Wikipedia article](https://en.wikipedia.org/wiki/Flexible_array_member):
> It is common to allocate sizeof(struct) + array_len*sizeof(array element) bytes.
> 
> This is not wrong, but it may allocate a few more bytes than necessary: the compiler may be re-purposing some of the padding that is included in sizeof(struct). Should this be a concern, macros are available[3] to compute the minimum size while ensuring that the compiler's padding is not disrupted.
> 
> As the array may start in the padding before the end of the structure, its content should always be accessed via indexing (arr[i]) or offsetof, not sizeof.

So I should expect such layout:
```
[main struct][keyfiles_length][blob with keyfiles]
             ^ offsetof(struct cust_salt, keyfiles_length)
```

I use `keyfiles_length[]` through field. So it may start earlier than `sizeof()` if `offsetof` and `sizeof` are different. Then using max offset for the blob leaves uninitialized memory between `keyfiles_length[]` and the blob. But this memory is hashed by `fmt_default_dyna_salt_hash`. So salts valid for dupe-removal might or might not merge depending on contents of memory.

I should use `mem_calloc` instead of `mem_alloc` to allocate the salt.

But I implemented `kpool` approach already and I like it.

AlekseyCherepanov commented:
I mean my actual use would look the following way in case `offsetof != sizeof`:
```
               v sizeof         vv not initialized
(main struct[  )keyfiles_length]  [blob with keyfiles]
            ^ offsetof            ^ sizeof + size of keyfiles_length[]
```

AlekseyCherepanov commented:
I rewrote commit reverting back from `dyna_salt`. I replaced keyfiles' data with precomputed `kpool`. So now there is additional 64 bytes in the salt structure that are not used without keyfiles.

> I think order of keyfiles does not affect result, so after more experiments I am going to add sorting to allow dupe-removal for such ciphertexts.

Precomputed `kpool` in salt means that ciphertexts with the same keyfiles in different order would be matched as identical.

Also I set `nkeyfiles` to 1 in the end. `nkeyfiles` is used as boolean value in other places. There can be sets of keyfiles that produce identical `kpool` contents. For instance, empty keyfile does not change `kpool` contents, so it can be added to any set. (Though such example should be unlikely in real life.)

I moved `keyfiles -> kpool` conversion from `apply_keyfiles` to `get_salt`. So it should speed up cracking.

New code can crack `TC-Normal-Volume` from `john-samples`. Normal and hidden volumes can be cracked in one session. So it solves #5072 for cpu formats.

I could not crack `TC-STD-Volume` sample. But older version without my changes could not crack it too. I'll solve it later.

This PR is ready. Thanks!

solardiz reviewed (approved):
I cannot properly review this without looking into the underlying issues myself first, but nothing in here looks necessarily wrong and I guess @AlekseyCherepanov knows what he's doing here.

@magnumripper Do we just merge this PR, or do you want to review first?

magnumripper commented:
> I cannot properly review this without looking into the underlying issues myself first, but nothing in here looks necessarily wrong and I guess @AlekseyCherepanov knows what he's doing here.
> 
> @magnumripper Do we just merge this PR, or do you want to review first?

Let's trust Aleksey
